### ISSUE TITLE
[QUESTION] Conan as a build system

### ISSUE THREAD

**Issue by acmeijer** [conan-io/conan#5296]
I have a question about the workflow when it comes to compiling C++ libraries and creating conan packages from these. 

Here is my current situation:
1. I define how my library should compile in CMake
2. If I have dependencies I can specify these in the requirements() part of my conanfile.py
3. To create the package I can run 'conan create' which relies on me having specified how to compile my library in a build() function. This function is also required when my conan package is rebuilt from sources.
4. I am using the CLion IDE for development

Now if I am developing my library itself within CLion then I am using conan to install my dependencies and after that CMake is in control of the build, so my steps are:
1. `conan install` (can use the plugin for this)
2. `cmake -DMYOPTION=MYVALUE` (set in CLion configuration or on command line)
3.  `make`

Now if I want to create my package, I am instead relying on how my build has been defined in the build() method of the conanfile.py. Ideally this matches the above steps, but this might not be the case exactly and in any case requires manual synchronisation - in the above example I would have had to write `cmake.definitions['MYOPTION'] = 'MYVALUE'`.

The solution then seems to be to always use 'conan build' and rely on the build() method as the single source of truth but there is no CLion support for this kind of workflow (yet?). 

My question is whether this is an accurate understanding of the situation and whether there is some advice as to how to avoid the above issue?

**Comment by danimtb** [comment]
Hi @acmeijer and thanks a lot for your question!

Indeed this is a thing we are aware of. As you mention, the way CMake is called from CLion versus the one used in the *conanfile.py* is different right now if you are setting the options from one side and want to use them in the other.

We call this issue the problem "Build Helpers VS. Generators", as in CMake for example, the generator is performing some checks that are useful when calling CMake manually, however the CMake helper is also setting some flags that will be missing the same way when you call CMake manually too.

The problem here is similar to CLion using different CMake flags than the recipe. We envision the solution to this problem with a feature in which Conan could provide users not only with a file with the information about dependencies (`conanbuilinfo.cmake`) but with a "toolchain file" with the flags the build helper sets. That could be injected to CMake so the flags are available no matter how or from you are executing CMake.

This "toolchain file" could be used by the CLion plugin so the same flags are used. In this case, the flags will be adjusted always from the recipe as it is the single source of truth for every build.

Does this make sense? Will it work for you?

I have added this issue to a project where we are gathering all the issues related to this "Build Helpers VS. Generators" problem. You can check them also find similar ones.

Thanks!

**Comment by acmeijer** [comment]
Hey @danimtb!

Thanks for your quick reply and great to know you are aware of this. I think the proposed solution sounds like a good one, however I am still curious how this works from the build() function in the conanfile, so if I understand it correctly:

the build() part of the conanfile is parsed and is used to create the toolchain file that you mentioned. You then simply include() it in your CMakeLists.txt and you will get the same build as when using the build() part of the conanfile. However what happens if conan itself is actually triggering the build? Would you not then have a cmake build where the toolchain is defined twice? I.e. once in the build() part of the conanfile and then again once cmake is invoked which includes the toolchain file?

If that is the case I guess there will need to be a mechanism that prevents the toolchain file being included if some variable is defined...

**Comment by danimtb** [comment]
Well, the idea is still not under development but I see two ways of achieving this:

1. The generators are the ones in charge of creating the toolchain files and then the build helper will just inject that file.

2. Creating a new concept like "toolchain generators" to differentiate this from the dependency generators.

3. Having a ``pre_build()`` step in the recipe where CMake defintions are set and the build helper is responsible for generating the "toolchain file" that will be used later in the ``build()`` step

The manual usage or IDE usage of that file will be something like ``cmake -DTOOLCHAIN_FILE=conancmaketoolchain.cmake`` (Or use ``-DCMAKE_PROJECT_<PROJECT-NAME>_INCLUDE`` as tested [here](https://github.com/conan-io/conan/issues/2956#issuecomment-393841871)).

Of course, we have to think about all the possibilities to make it work smoothly from ``conan build`` command, plain CMake commands or IDEs, but I think you get a general idea.

**Comment by acmeijer** [comment]
Yes I see and I think it makes sense. It sounds like it will make the boundary between where CMake and Conan interact much cleaner.

I'd be happy to help out e.g. by testing out changes in our workflow.

Would you like me to close this issue?

**Comment by danimtb** [comment]
Let's keep it open so we can consider this for the feature development. Moreover, the integration of the described above should also fit with IDEs plugins.

Thanks for your thoughts @acmeijer

**Comment by memsharded** [comment]
I am closing this ticket as solved long time ago:

- New CMake integrations ``CMakeToolchain``, ``CMakeDeps``, that achieve the exact same result when using ``conan build`` than when using ``conan install`` + ``cmake ..``
- New generation of CMake presets that further streamline the process and integration
- New Conan 2.0 integration with CMake via CMake dependency providers with ``cmake-conan``
- CLion new plugin using the 2.0 ``cmake-conan`` integration.
